src/importsource.cpp

```diff
@@ -69,7 +70,7 @@ void ImportSource::setModel(const ModelPtr &model)
 
 bool ImportSource::hasModel() const
 {
-    return mPimpl->mModel != nullptr;
+    return !mPimpl->mModel.expired();
```

Comment:
Will this fail when mModel is nullptr? Or, how should we clear the model from here (previously set it to nullptr)?

Reply:
`mModel` is now a weak pointer, so you cannot assign `nullptr` to it. I haven't checked your use of `mModel`, but if you need it to be reset to `nullptr` (after it was set to something meaningful) then, you will need to find a way around it (a "diry" solution would be to use a `bool` to determine whether `mModel` is set).

My "fix" was merely to prevent the memory leak, which occurs because of circular referencing. So, either `mModel` needs to be a weak pointer, or the other struct/class member responsible for the circular reference.

Reply:
Yeah, we do need a way to unset it.  It's currently used in these functions below.  I'll do some more digging :)
```
void clearComponentImports(const ComponentPtr &component)
{
    if (component->isImport()) {
        component->importSource()->setModel(nullptr);
    }
    for (size_t c = 0; c < component->componentCount(); ++c) {
        if (component->component(c)->isImport()) {
            component->component(c)->importSource()->setModel(nullptr);
        }
    }
}

void Importer::clearImports(ModelPtr &model)
{
    // Clears the models from all import sources in the model.
    for (size_t u = 0; u < model->unitsCount(); ++u) {
        if (model->units(u)->isImport()) {
            model->units(u)->importSource()->setModel(nullptr);
        }
    }
    for (size_t c = 0; c < model->componentCount(); ++c) {
        clearComponentImports(model->component(c));
        if (model->component(c)->isImport()) {
            model->component(c)->importSource()->setModel(nullptr);
        }
    }
}
```

Reply:
Should be good now ... but do we need to add something into the python? I'm getting a failure of a test there?